cmd/opampsupervisor/e2e_test.go

```diff
@@ -0,0 +1,319 @@
+		shutdown()
+	})
+	return httpSrv.Listener.Addr().String(), connectedChan, send
+}
+
+func newSupervisor(t *testing.T, configType string, extraConfigData map[string]string) *supervisor.Supervisor {
+	wd, err := os.Getwd()
+	require.NoError(t, err)
+	tpl, err := os.ReadFile(path.Join(wd, "testdata", "supervisor", "supervisor_"+configType+".yaml"))
+	require.NoError(t, err)
+
+	templ, err := template.New("").Parse(string(tpl))
+	require.NoError(t, err)
+
+	var buf bytes.Buffer
+	var extension string
+	if runtime.GOOS == "windows" {
+		extension = ".exe"
+	}
+	configData := map[string]string{
+		"goos":      runtime.GOOS,
+		"goarch":    runtime.GOARCH,
+		"extension": extension,
+	}
+
+	for key, val := range extraConfigData {
+		configData[key] = val
+	}
+	err = templ.Execute(&buf, configData)
+	require.NoError(t, err)
+	cfgFile, _ := os.CreateTemp(t.TempDir(), "config_*.yaml")
+	_, err = cfgFile.Write(buf.Bytes())
+	require.NoError(t, err)
+
+	s, err := supervisor.NewSupervisor(zap.NewNop(), cfgFile.Name())
+	require.NoError(t, err)
+
+	return s
+}
+
+func TestSupervisorStartsCollectorWithRemoteConfig(t *testing.T) {
+	var agentConfig atomic.Value
+	addr, connected, send := newServer(
+		t,
+		defaultConnectingHandler,
+		server.ConnectionCallbacksStruct{
+			OnMessageFunc: func(_ types.Connection, message *protobufs.AgentToServer) *protobufs.ServerToAgent {
+				if message.EffectiveConfig != nil {
+					config := message.EffectiveConfig.ConfigMap.ConfigMap[""]
+					if config != nil {
+						agentConfig.Store(string(config.Body))
+					}
+				}
+
+				return &protobufs.ServerToAgent{}
+			},
+		})
+
+	s := newSupervisor(t, "basic", map[string]string{"url": addr})
+	defer s.Shutdown()
+
+	waitForConnection(connected, true)
+
+	cfg, hash, input, output := createSimplePipelineConfig(t)
+
+	send(&protobufs.ServerToAgent{
+		RemoteConfig: &protobufs.AgentRemoteConfig{
+			Config: &protobufs.AgentConfigMap{
+				ConfigMap: map[string]*protobufs.AgentConfigFile{
+					"": {Body: cfg.Bytes()},
+				},
+			},
+			ConfigHash: hash,
+		},
+	})
+
+	require.Eventually(t, func() bool {
+		cfg, ok := agentConfig.Load().(string)
+		if ok {
+			// The effective config may be structurally different compared to what was sent,
+			// so just check that it includes some strings we know to be unique to the remote config.
+			return strings.Contains(cfg, input.Name()) && strings.Contains(cfg, output.Name())
+		}
+
+		return false
+	}, 5*time.Second, 500*time.Millisecond, "Collector was not started with remote config")
+
+	n, err := input.WriteString("{\"body\":\"hello, world\"}\n")
+	require.NotZero(t, n, "Could not write to input file")
+	require.NoError(t, err)
+
+	require.Eventually(t, func() bool {
+		logRecord := make([]byte, 1024)
+		n, _ := output.Read(logRecord)
+
+		return n != 0
+	}, 10*time.Second, 500*time.Millisecond, "Log never appeared in output")
+}
+
+func TestSupervisorRestartsCollectorAfterBadConfig(t *testing.T) {
```

Comment:
I was not able to locate the part of the code in this test that is responsible for supplying the "bad" config. Can you add comments to help understand how the test works?

Reply:
Good catch, I was unintentionally relying on the fact that the Supervisor will start the Collector with a bad config if it isn't given a remote config. I've updated this to use a config that is known to be bad.